resource_work_time_from_contracts/models/resource_mixin.py

```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
+                or (
+                    attendance_from[0] == earliest_attendance[0]
+                    and attendance_from[1].hour_from < earliest_attendance[1].hour_from
+                )
+            ):
+                earliest_attendance = attendance_from
+            if attendance_to is not None and (
+                latest_attendance is None
+                or attendance_to[0] > latest_attendance[0]
+                or (
+                    attendance_to[0] == latest_attendance[0]
+                    and attendance_to[1].hour_to > latest_attendance[1].hour_to
+                )
+            ):
+                latest_attendance = attendance_to
+        if earliest_attendance is None:
+            return (None, None)
+        return (earliest_attendance[1], latest_attendance[1])
+
+    def get_calendar_for_date(self, date):
+        """
+        Get the resource.calendar of the contract that applies to the provided
+        date.
+        """
+        # there can be multiple contracts active at the same time. only the
+        # first one found will be used.
+        current_contracts = self._get_active_contracts(date, date)
+        if not current_contracts:
+            return None
+        return current_contracts[0].resource_calendar_id
+
+    def _sum_intervals(self, intervals):
+        result = defaultdict(float)
+        for start, stop, meta in intervals:
+            result[start.date()] += (stop - start).total_seconds() / 3600
+        return sorted(result.items())
+
+    def _get_round_day_round_func(self, rounding_mode, granularity):
+        def round_day_default(work_time, normal_work_time):
+            # this is the same rounding computation as in
+            # resource.resource_mixin.get_work_days_data().
+            return (
+                float_utils.round(ROUNDING_FACTOR * work_time / normal_work_time)
+                / ROUNDING_FACTOR
+            )
+
+        def round_day_round(work_time, normal_work_time):
+            return (
+                float_utils.round((work_time / normal_work_time) / granularity)
+                * granularity
+            )
+
+        def round_day_ceil(work_time, normal_work_time):
+            return (
+                math.ceil(
+                    (work_time / normal_work_time) / granularity
+                    - granularity * DAY_ROUNDING_CEIL_THRESHOLD
+                )
+                * granularity
+            )
+
+        if rounding_mode == DAY_ROUNDING_MODE_ROUND:
+            return round_day_round
+        if rounding_mode == DAY_ROUNDING_MODE_CEIL:
+            return round_day_ceil
+        return round_day_default
+
+    def _get_day_rounding_func(self):
+        ir_config_parameter_model = self.env["ir.config_parameter"].sudo()
+        rounding_mode = ir_config_parameter_model.get_param(DAY_ROUNDING_MODE_PARAM)
+        rounding_granularity = ir_config_parameter_model.get_param(
+            DAY_ROUNDING_GRANULARITY_PARAM
+        )
```

Comment:
Do these come with defaults? We should either fall back to a default, or populate these params by default, or both.

Reply:
yes, all of these has defaults. `ir.config_parameter.get_param()` return `False` if the parameter is not defined. in that case, we use:

*   `DAY_ROUNDING_MODE_PARAM`: `round_day_default()`
*   `DAY_ROUNDING_GRANULARITY_PARAM`: `DEFAULT_DAY_ROUNDING_GRANULARITY` (`1`)